shell/browser/badging/badge_manager.cc

```diff
@@ -0,0 +1,81 @@
+// Copyright 2018 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "shell/browser/badging/badge_manager.h"
+
+#include <tuple>
+#include <utility>
+
+#include "base/i18n/number_formatting.h"
+#include "base/strings/utf_string_conversions.h"
+#include "build/build_config.h"
+#include "content/public/browser/browser_task_traits.h"
+#include "content/public/browser/browser_thread.h"
+#include "content/public/browser/render_frame_host.h"
+#include "content/public/browser/render_process_host.h"
+#include "content/public/browser/web_contents.h"
+#include "shell/browser/badging/badge_manager_factory.h"
+#include "shell/browser/browser.h"
+#include "ui/base/l10n/l10n_util.h"
+#include "ui/strings/grit/ui_strings.h"
+
+namespace badging {
+
+BadgeManager::BadgeManager() = default;
+BadgeManager::~BadgeManager() = default;
+
+// static
+void BadgeManager::BindFrameReceiver(
+    content::RenderFrameHost* frame,
+    mojo::PendingReceiver<blink::mojom::BadgeService> receiver) {
+  DCHECK_CURRENTLY_ON(content::BrowserThread::UI);
+
+  auto* browser_context =
+      content::WebContents::FromRenderFrameHost(frame)->GetBrowserContext();
+
+  auto* badge_manager =
+      badging::BadgeManagerFactory::GetInstance()->GetForBrowserContext(
+          browser_context);
+  if (!badge_manager)
+    return;
+
+  auto context = std::make_unique<FrameBindingContext>(
+      frame->GetProcess()->GetID(), frame->GetRoutingID());
+
+  badge_manager->receivers_.Add(badge_manager, std::move(receiver),
+                                std::move(context));
+}
+
+std::string BadgeManager::GetBadgeString(base::Optional<int> badge_content) {
+  if (!badge_content)
+    return "•";
+
+  if (badge_content > kMaxBadgeContent) {
+    return base::UTF16ToUTF8(l10n_util::GetStringFUTF16(
+        IDS_SATURATED_BADGE_CONTENT, base::FormatNumber(kMaxBadgeContent)));
+  }
+
+  return base::UTF16ToUTF8(base::FormatNumber(badge_content.value()));
+}
+
+void BadgeManager::SetBadge(blink::mojom::BadgeValuePtr mojo_value) {
+  if (mojo_value->is_number() && mojo_value->get_number() == 0) {
+    mojo::ReportBadMessage(
+        "|value| should not be zero when it is |number| (ClearBadge should be "
+        "called instead)!");
+    return;
+  }
+
+  base::Optional<int> value =
+      mojo_value->is_flag() ? base::nullopt
+                            : base::make_optional(mojo_value->get_number());
+
+  electron::Browser::Get()->SetBadgeCount(value);
```

Comment:
Should `navigator.setAppBadge` only set badge for the source window instead of all windows in the app?

Reply:
@zcbenz I'm not sure I understand what you are asking here? `navigator.setAppBadge` is not window specific but for the whole app, therefore, I just tied it into [app.setBadgeCount(count)](https://www.electronjs.org/docs/api/app#appsetbadgecountcount-linux-macos).  Am I missing something?

Reply:
On Windows the badge icons are set per-window, and this PR's implementation of `app.setBadgeCount` sets badge icon for all current windows to emulate the app-wide badge icons on macOS.

But to emulate app-wide badge icons we also need to automatically set badge icon for new windows, otherwise the new windows won't have the badges:

```js
app.setBadgeCount(1)
new BrowserWindow()  // window has no badge
```

So to correctly implement `app.setBadgeCount` on Windows we would end up with a rather complex system. And in future if we decide to add a per-window `BrowserWindow.setBadgeCount` API for Windows, this will get more complicated.

I think we don't really need to implement `app.setBadgeCount` on Windows, we can just add an `BrowserWindow.setBadgeCount` API and let users decide which one to use. This would also align with existing APIs, e.g. `app.dock.setIcon` and `BrowserWindow.setOverlayIcon`.

For the `navigator.setAppBadge` API, while it claims to set app badge, but it actually only works in PWA apps that run in standalone windows, so on Windows this API essentially sets badge for current window instead of all windows in the browser:
https://github.com/w3c/badging/blob/master/explainer.md#ux-treatment

Reply:
@zcbenz ok I understand what you are asking.  I can change the code to set the badge for all existing windows.  Since the main reason for this PR is to support navigator.setAppBadge/clearAppBadge I could just make that work in Windows and not add Windows support to app.get/setBadgeCount.  Another option would be to document in our documentation that on Windows app.setBadgeCount will only affect existing windows.  What do you think?

Reply:
I prefer not adding `app.get/setBadgeCount` APIs for Windows, setting same badge for all windows is a weird behavior and I don't think there is any app doing that.

Reply:
@zcbenz I removed the changes to app.get/setBadgeCount for Windows, so this PR is now just a fix to make navigator.setAppBadge/clearAppBadge work in Electron.